flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2114,6 +2117,93 @@ class ScalarExprLowering {
         });
   }
 
+  /// Helper structure to track potential copy-in of non contiguous variable
+  /// argument into a contiguous temp. It is used to deallocate the temp that
+  /// may have been created as well as to the copy-out from the temp to the
+  /// variable after the call.
+  struct CopyOutPair {
+    ExtValue var;
+    ExtValue temp;
+    // Flag to indicate if the argument may have been modified by the
+    // callee, in which case it must be copied-out to the variable.
+    bool argMayBeModifiedByCall;
+    // Optional boolean value that, if present and false, prevents
+    // the copy-out and temp deallocation.
+    llvm::Optional<mlir::Value> restrictCopyAndFreeAtRuntime;
+  };
+  using CopyOutPairs = llvm::SmallVector<CopyOutPair, 4>;
+
+  /// Helper to read any fir::BoxValue into fir::UnboxedValue, or
+  /// fir::CharArrayBoxValue. This will lose any non contiguous stride
```

Comment:
What about arrays of other intrinsic types or scalar characters? Arrays should really be ArrayBoxValue.